gh-100228: Warn from os.fork() if other threads exist.

gpshead opened the pull request:
Not comprehensive, best effort warning. There are cases when threads exist that this code currently does not detect.

Starting with a `DeprecationWarning` for now, it is less disruptive than something like `RuntimeWarning` and most likely to only be seen in people's CI tests - a good place to start with this messaging.

TODO:
 * [x] doctests for multiprocessing failed with an assertion crash in [2854473](https://github.com/python/cpython/pull/100229/commits/28544731190331bebae1b135b3fd1abb4b835a37) about the error being set unexpectedly on one later commit, why? did I fix that? I need to figure out how to run those and try to reproduce it. _(could not reproduce, no way it could be related seen upon code inspection)_

<!-- gh-issue-number: gh-100228 -->
* Issue: gh-100228
<!-- /gh-issue-number -->

gpshead committed:
Warn from os.fork() if other threads exist.

Not comprehensive, best effort.

gpshead committed:
fix existing tests for the new warning.

gpshead committed:
static

gpshead committed:
&_Py_ID() for _active and _limbo.

gpshead committed:
news wording tweak

gpshead committed:
Add platform API calls on Linux & Mac, polish up.

* Only warn in the parent process.
* Add a test to test_os that a non-Python thread triggers the warning.
* Avoid the system calls to count threads if we've already warned once.
* The macOS API calls were written blind based on docs, will fix later
  if CI fails.

gpshead committed:
Merge branch 'warn/os.fork/threads' of github.com:gpshead/cpython into warn/os.fork/threads

gpshead committed:
Probably fix the macOS crashes.

I don't want the array of thread info but it may insist?

Yhg1s commented on Modules/posixmodule.c:
```diff
@@ -6737,6 +6739,96 @@ os_register_at_fork_impl(PyObject *module, PyObject *before,
+            Py_XDECREF(threading_active);
+            return;
+        }
+        // Worst case if someone replaced threading._active or threading._limbo
```
If it's not fatal that this happens, these shouldn't be asserts. Otherwise code that works fine will start aborting for the wrong reason when run with a debug build (or, for example, when imported in a build system that enables asserts when running tests *cough*).

gpshead replied:
asserts removed.

Yhg1s commented on Modules/posixmodule.c:
```diff
@@ -6737,6 +6739,96 @@ os_register_at_fork_impl(PyObject *module, PyObject *before,
 
+// Common code to raise a warning if we know there is more than one thread
+// running in the process.  Best effort, silent if unable to count threads.
+// Constraint: Avoids locks. Quick. Never leaves an error set.
```
The 'slow path' doesn't avoid locks, as it may release the GIL in the getmodule/getattr calls. If it's a concern perhaps we could expose the 'known number of threads' information from the runtime?

gpshead replied:
It's not a concern, reworded.

Yhg1s commented on Modules/posixmodule.c:
```diff
@@ -6737,6 +6739,96 @@ os_register_at_fork_impl(PyObject *module, PyObject *before,
+            num_python_threads = atoi(field);  // 0 on error
+        }
+    }
+#endif
```
Is "figure out the Windows way" on the TODO list, or already known not to work well/fast enough?

gpshead replied:
this one is easy: Windows isn't POSIX and does not have fork() so it'll never end up here.

Yhg1s commented on Lib/test/test_os.py:
```diff
@@ -4534,6 +4534,31 @@ def test_fork(self):
+                    warnings.filterwarnings(
+                            "always", category=DeprecationWarning)
+                    if os.fork() == 0:
+                        assert len(ws) == 0  # Warn only the parent
```
Oh yay tests that won't work right when PYTHONOPTIMIZE is set :( Since it's already done this way in other tests I won't suggest fixing it, but uugh.

gpshead replied:
I cleaned it up a bit anyways. assert_python_ok doesn't appear to reconstruct -O flags on the command line, but at least explicitly having it set PYTHONOPTIMIZE=0 means the asserts should always work.

there was nothing checking for a child process error anyways, it could've gotten lost. fixed now.

Yhg1s commented on Misc/NEWS.d/next/Library/2022-12-13-17-29-09.gh-issue-100228.bgtzMV.rst:
```diff
@@ -0,0 +1,4 @@
+A :exc:`DeprecationWarning` may be raised when :func:`os.fork()` or
+:func:`os.forkpty()` is called from multi-threaded processes. This is to
+help people realize they are doing something unsafe. Lack of a warning
+does not indicate that the fork call was actually safe.
```
```suggestion
:func:`os.forkpty()` is called from multi-threaded processes.  Forking
with threads is unsafe and can cause deadlocks, crashes and subtle
problems. Lack of a warning does not indicate that the fork call was
actually safe, as Python may not be aware of all threads.
```

Yhg1s commented on Modules/_testcapimodule.c:
```diff
@@ -870,6 +873,45 @@ test_thread_state(PyObject *self, PyObject *args)
+}
+
+// These can be used to test things that care about the existence of another
+// thread that the Python runtime doesn't know about.
```
```suggestion
// thread that the threading module doesn't know about.
```

The Python runtime does know about these threads (they hold the GIL even :). It would be kinda nice if we could tell whether any (non-main) PyThreadStates were active, since that would detect these kinds of threads too. I guess we want to be careful not to warn when the extra threads have recently exited and _right now_ there is only one active thread.

gpshead replied:
The thread this pair launches and tells to exit never touches the GIL.

The PyThread APIs are not APIs that use the GIL, they're low level CPython internal only things that merely abstract platform thread and lock calls, which will only ever be pthread_create and friends given my use is POSIX specific. Using them avoided me needing to figure those out and possibly carry more ifdef conditionals in here.

I suppose that means someone could use these in a non-posix test such as on Windows for another purpose if desired. If we ever refactored this into a Python API to return the current number of threads in the process, that'd presumably want to use it.

Yhg1s commented on Modules/posixmodule.c:
```diff
@@ -6737,6 +6739,96 @@ os_register_at_fork_impl(PyObject *module, PyObject *before,
+    }
+    if (num_python_threads > 1) {
+        PyErr_WarnFormat(
+                PyExc_DeprecationWarning, 1,
```
FWIW I think DeprecationWarning is the right call for now. It means it will be invisible unless explicitly enabled, and gives us time to evaluate the thread-detection in the field, as it were, while still giving people the tool they need to detect and avoid this particular problem.

Yhg1s commented on Modules/posixmodule.c:
```diff
@@ -6737,6 +6739,96 @@ os_register_at_fork_impl(PyObject *module, PyObject *before,
+    if (num_python_threads > 1) {
+        PyErr_WarnFormat(
+                PyExc_DeprecationWarning, 1,
+                "multi-threaded process, %s() may cause deadlocks.", name);
```
Should we point to more information to understand this issue somewhere, like a FAQ entry? I know we don't usually do this but this is somewhat of an esoteric issue, might be deep inside a library (e.g. something using multiprocessing), and we _have_ been trying to improve error messages to be more helpful...

gpshead replied:
I don't have a good canonical link to propose right now. The other thing that often happens is that the error message winds up with a Google search result of a stackoverflow or similar live content user moderated question linking to the current best sources of info.

For multiprocessing, I intend to do the warning about start_method `fork` as the default being used. In that specific default-fork situation where multiprocessing will provide its own DeprecationWarning, it probably makes sense to have multiprocessing silence this one on its call.

gpshead replied:
Wording improved to include the pid and indicate that deadlocks could happen in the child process. Searching, I don't see a canonical online reference to the fork+threads issue of high enough quality and domain owner stability that I'd want to codify it within our error message. So I'll leave this without a URL for now, we can add one later if desired.

If we want one, we probably need to write a little section covering the topic in our docs somewhere and link to that.

gpshead committed:
Fix test_os race condition.

Let the thread free and clear things as it ends after we unblock it.

gpshead committed:
remove the already_warned logic.

It could hide more than we intend and wouldn't do what I wanted in child
processes that inherit the parents warnings state anyways.

gpshead committed:
Improve the warning message.

gpshead committed:
Update Misc/NEWS.d/next/Library/2022-12-13-17-29-09.gh-issue-100228.bgtzMV.rst

Co-authored-by: T. Wouters <thomas@python.org>

gpshead committed:
Update a comment

Co-authored-by: T. Wouters <thomas@python.org>

gpshead committed:
Clean up comments; add a final PyErr_Clear warning

gpshead committed:
Improve test_os robustness.

gpshead committed:
ignore expected warnings.

gpshead committed:
improve error message.

gpshead committed:
silence the warning in more tests.

gpshead committed:
line length fixes, comments++, DECREF vs XDECREF

gpshead committed:
Merge branch 'main' into warn/os.fork/threads
